File: day2/aoc2a.py
def main():
    """MAIN"""
    max_red = 12
    max_green = 13
    max_blue = 14

    count = 0

    with open("input.txt", "r", encoding="UTF-8") as file:
        for line in file:
            print(line.strip())
            id_str, rest = line.strip().split(":")
            game_id = int(id_str.split(" ")[1])
            runs = [
                [y.strip().split(" ") for y in x.strip().split(",")]
                for x in rest.split(";")
            ]
            dead = 0
            for run in runs:
                if dead:
                    break
                for color in run:
                    if dead:
                        break
                    if color[1] == "red":
                        if int(color[0]) > max_red:
                            dead = 1
                    elif color[1] == "green":
                        if int(color[0]) > max_green:
                            dead = 1
                    elif color[1] == "blue":
                        if int(color[0]) > max_blue:
                            dead = 1
            if dead:
                print(line.strip(), " is bad")
            else:
                count += game_id
    print(count)


if __name__ == "__main__":
    main()

File: day2/aoc2a.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

char **split(char *input, char *separator, int *len)
{
    char **result = 0;
    char tmp[strlen(input)];
    strcpy(tmp, input);
    size_t count = 0;
    char *token = strtok(tmp, separator);
    while (token)
    {
        token = strtok(NULL, separator);
        count++;
    }

    result = malloc(sizeof(char *) * count);
    *len = count;

    count = 0;
    token = strtok(input, separator);
    result[count] = token;
    while (token)
    {
        count++;
        token = strtok(NULL, separator);
        result[count] = token;
    }
    return result;
}

int main(int argc, char **argv)
{
    const int max_red = 12;
    const int max_green = 13;
    const int max_blue = 14;

    char *input_file = "inputa.txt";
    FILE *fptr;
    if ((fptr = fopen(input_file, "r")) == NULL)
    {
        printf("Error! opening file");
        // Program exits if the file pointer returns NULL.
        exit(1);
    }

    const unsigned MAX_LENGTH = 256;
    char buffer[MAX_LENGTH];

    long unsigned int count = 0;
    int test = 1;
    int sum_of_ids = 0;

    while (fgets(buffer, MAX_LENGTH, fptr))
    {
        printf("%d\n", test);
        test++;
        int game_id;
        printf("%s", buffer);
        char game_id_buff[MAX_LENGTH];
        strcpy(game_id_buff, buffer);

        // get game id
        int count = 0;
        char **splitted = split(game_id_buff, " ", &count);
        char game_id_temp[5];
        strncpy(game_id_temp, splitted[1], strlen(game_id_buff) - 2);
        game_id = atoi(game_id_temp);
        printf("Game ID = %d\n", game_id);

        for (int i = 2; i < count; i++)
        {
            printf("%s\n", splitted[i]);
        }

        free(splitted);
    }

    fclose(fptr);
    return 0;
}
